File: src/BehaviorsSDKNative/Microsoft.Xaml.Interactions/InvokeCommandAction.cpp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
#include "pch.h"
#include "InvokeCommandAction.h"

using namespace Platform;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Input;
using namespace Microsoft::Xaml::Interactivity;
using namespace Microsoft::Xaml::Interactions::Core;

namespace DependencyProperties
{
	static DependencyProperty^ Command = DependencyProperty::Register(
		"Command",
		ICommand::typeid,
		InvokeCommandAction::typeid,
		ref new PropertyMetadata(nullptr));

	static DependencyProperty^ CommandParameter = DependencyProperty::Register(
		"CommandParameter",
		Object::typeid,
		InvokeCommandAction::typeid,
		ref new PropertyMetadata(nullptr));

	static DependencyProperty^ InputConverter = DependencyProperty::Register(
		"InputConverter",
		IValueConverter::typeid,
		InvokeCommandAction::typeid,
		ref new PropertyMetadata(nullptr));

	static DependencyProperty^ InputConverterParameter = DependencyProperty::Register(
		"InputConverterParameter",
		Object::typeid,
		InvokeCommandAction::typeid,
		ref new PropertyMetadata(nullptr));

	static DependencyProperty^ InputConverterLanguage = DependencyProperty::Register(
		"InputConverterLanguage",
		String::typeid,
		InvokeCommandAction::typeid,
		ref new PropertyMetadata(""));
}

DependencyProperty^ InvokeCommandAction::CommandProperty::get()
{
	return DependencyProperties::Command;
}

DependencyProperty^ InvokeCommandAction::CommandParameterProperty::get()
{
	return DependencyProperties::CommandParameter;
}

DependencyProperty^ InvokeCommandAction::InputConverterProperty::get()
{
	return DependencyProperties::InputConverter;
}

DependencyProperty^ InvokeCommandAction::InputConverterParameterProperty::get()
{
	return DependencyProperties::InputConverterParameter;
}

DependencyProperty^ InvokeCommandAction::InputConverterLanguageProperty::get()
{
	return DependencyProperties::InputConverterLanguage;
}

Object^ InvokeCommandAction::Execute(Object^ sender, Object^ parameter)
{
	if (this->Command == nullptr)
	{
		return false;
	}

	Object^ resolvedParamter;
	if (this->CommandParameter != nullptr)
	{
		resolvedParamter = this->CommandParameter;
	}
	else if (this->InputConverter != nullptr)
	{
		resolvedParamter = this->InputConverter->Convert(
			parameter,
			Object::GetType(),
			this->InputConverterParameter,
			this->InputConverterLanguage);
	}
	else
	{
		resolvedParamter = parameter;
	}

	if (!this->Command->CanExecute(resolvedParamter))
	{
		return false;
	}

	this->Command->Execute(resolvedParamter);
	return true;
}

File: src/BehaviorsSDKManaged/Microsoft.Xaml.Interactivity.Shared/Core/InvokeCommandAction.cs
﻿// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Diagnostics.CodeAnalysis;

using System.Windows.Input;

#if WinUI
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Data;
#else
using Windows.UI.Xaml;
using Windows.UI.Xaml.Data;
#endif

namespace Microsoft.Xaml.Interactivity;

/// <summary>
/// Executes a specified <see cref="global::System.Windows.Input.ICommand"/> when invoked. 
/// </summary>
public sealed class InvokeCommandAction : DependencyObject, IAction
{
    /// <summary>
    /// Identifies the <seealso cref="Command"/> dependency property.
    /// </summary>
    [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
    public static readonly DependencyProperty CommandProperty = DependencyProperty.Register(
        "Command",
        typeof(ICommand),
        typeof(InvokeCommandAction),
        new PropertyMetadata(null));

    /// <summary>
    /// Identifies the <seealso cref="CommandParameter"/> dependency property.
    /// </summary>
    [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
    public static readonly DependencyProperty CommandParameterProperty = DependencyProperty.Register(
        "CommandParameter",
        typeof(object),
        typeof(InvokeCommandAction),
        new PropertyMetadata(null));

    /// <summary>
    /// Identifies the <seealso cref="InputConverter"/> dependency property.
    /// </summary>
    [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
    public static readonly DependencyProperty InputConverterProperty = DependencyProperty.Register(
        "InputConverter",
        typeof(IValueConverter),
        typeof(InvokeCommandAction),
        new PropertyMetadata(null));

    /// <summary>
    /// Identifies the <seealso cref="InputConverterParameter"/> dependency property.
    /// </summary>
    [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
    public static readonly DependencyProperty InputConverterParameterProperty = DependencyProperty.Register(
        "InputConverterParameter",
        typeof(object),
        typeof(InvokeCommandAction),
        new PropertyMetadata(null));

    /// <summary>
    /// Identifies the <seealso cref="InputConverterLanguage"/> dependency property.
    /// </summary>
    [SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
    public static readonly DependencyProperty InputConverterLanguageProperty = DependencyProperty.Register(
        "InputConverterLanguage",
        typeof(string),
        typeof(InvokeCommandAction),
        new PropertyMetadata(string.Empty)); // Empty string means the invariant culture.

    /// <summary>
    /// Gets or sets the command this action should invoke. This is a dependency property.
    /// </summary>
    public ICommand Command
    {
        get
        {
            return (ICommand)this.GetValue(InvokeCommandAction.CommandProperty);
        }
        set
        {
            this.SetValue(InvokeCommandAction.CommandProperty, value);
        }
    }

    /// <summary>
    /// Gets or sets the parameter that is passed to <see cref="global::System.Windows.Input.ICommand.Execute(object)"/>.
    /// If this is not set, the parameter from the <seealso cref="Execute(object, object)"/> method will be used.
    /// This is an optional dependency property.
    /// </summary>
    public object CommandParameter
    {
        get
        {
            return this.GetValue(InvokeCommandAction.CommandParameterProperty);
        }
        set
        {
            this.SetValue(InvokeCommandAction.CommandParameterProperty, value);
        }
    }

    /// <summary>
    /// Gets or sets the converter that is run on the parameter from the <seealso cref="Execute(object, object)"/> method.
    /// This is an optional dependency property.
    /// </summary>
    public IValueConverter InputConverter
    {
        get
        {
            return (IValueConverter)this.GetValue(InvokeCommandAction.InputConverterProperty);
        }
        set
        {
            this.SetValue(InvokeCommandAction.InputConverterProperty, value);
        }
    }

    /// <summary>
    /// Gets or sets the parameter that is passed to the <see cref="IValueConverter.Convert"/>
    /// method of <see cref="InputConverter"/>.
    /// This is an optional dependency property.
    /// </summary>
    public object InputConverterParameter
    {
        get
        {
            return this.GetValue(InvokeCommandAction.InputConverterParameterProperty);
        }
        set
        {
            this.SetValue(InvokeCommandAction.InputConverterParameterProperty, value);
        }
    }

    /// <summary>
    /// Gets or sets the language that is passed to the <see cref="IValueConverter.Convert"/>
    /// method of <see cref="InputConverter"/>.
    /// This is an optional dependency property.
    /// </summary>
    public string InputConverterLanguage
    {
        get
        {
            return (string)this.GetValue(InvokeCommandAction.InputConverterLanguageProperty);
        }
        set
        {
            this.SetValue(InvokeCommandAction.InputConverterLanguageProperty, value);
        }
    }

    /// <summary>
    /// Executes the action.
    /// </summary>
    /// <param name="sender">The <see cref="global::System.Object"/> that is passed to the action by the behavior. Generally this is <seealso cref="Microsoft.Xaml.Interactivity.IBehavior.AssociatedObject"/> or a target object.</param>
    /// <param name="parameter">The value of this parameter is determined by the caller.</param>
    /// <returns>True if the command is successfully executed; else false.</returns>
    public object Execute(object sender, object parameter)
    {
        if (this.Command == null)
        {
            return false;
        }

        object resolvedParameter;
        if (this.ReadLocalValue(InvokeCommandAction.CommandParameterProperty) != DependencyProperty.UnsetValue)
        {
            resolvedParameter = this.CommandParameter;
        }
        else if (this.InputConverter != null)
        {
            resolvedParameter = this.InputConverter.Convert(
                parameter,
                typeof(object),
                this.InputConverterParameter,
                this.InputConverterLanguage);
        }
        else
        {
            resolvedParameter = parameter;
        }

        if (!this.Command.CanExecute(resolvedParameter))
        {
            return false;
        }

        this.Command.Execute(resolvedParameter);
        return true;
    }
}
